Some more automation on grading/steps/stats.py, and more tests

stefano-maggiolo opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/940)
<!-- Reviewable:end -->

lerks commented:
Reviewed 2 of 2 files at r2, 3 of 3 files at r3, 3 of 3 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cms/grading/steps/stats.py, line 63 at r3](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADLKDNADREIiBobSRJ:-LADLKDNADREIiBobSRK:b-8puuf4) ([raw file](https://github.com/cms-dev/cms/blob/9c5649a33b079a2ad6e44e62e0c9f27bb1a4a0ba/cms/grading/steps/stats.py#L63)):*
> ```Python
>     The first input stats can be None, in which case the second stats is copied
>     to the output (useful to treat the first merge of a sequence in the same
>     way as the others).
> ```

Could be made a bit more general by saying that if either is None we return the other, and if both are None we return None. Though I'm not saying I prefer it this way, just wanting to point this out.

---

*[cms/grading/steps/stats.py, line 117 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOdrK6tKauUWi3pIR:-LADOdrK6tKauUWi3pIS:bawpekn) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L117)):*
> ```Python
> 
>     for f in ["stdout", "stderr"]:
>         ret[f] = "\n===\n".join(d[f] for d in [ret, second_stats] if f in d)
> ```

Doing the join this way (pairwise) brings the overall complexity to N^2, whereas before it was N.

---

*[cms/grading/steps/stats.py, line 119 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOm698nrYwoTRCr59:-LADOm698nrYwoTRCr5A:b-qcvj68) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L119)):*
> ```Python
>         ret[f] = "\n===\n".join(d[f] for d in [ret, second_stats] if f in d)
>         if ret[f] == "":
>             del ret[f]
> ```

I'm not sure I like this. I'm imagining the following scenario: three commands, all produce stdout but only the last one produces stderr. The merged stdout will be `a\n===\nb\n===\nc`, the stderr will be `foo`. It's impossible to figure out which command produced the stderr. If it were instead `\n===\n\n===\nfoo` it would be uglier but clear.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 32 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADQkTw3iJ61qtuqgwV:-LADQkTw3iJ61qtuqgwW:brhumzt) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L32)):*
> ```Python
> import unittest
> 
> from cmstestsuite.unit_tests.grading.steps.fakeisolatesandbox \
> ```

Same here.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 63 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADRTz33kRKUBjWtoMD:-LADRTz33kRKUBjWtoME:b721erh) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L63)):*
> ```Python
>         # Stdout and stderr are encoded in UTF-8.
>         self.assertEqual(stats["stdout"], "o")
>         self.assertEqual(stats["stderr"], "你好")
> ```

Did you consider importing `get_stats` from `stats_test` (or copying it over) to simplify this and all the ones below?

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 68 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZ6Ca0C46gDAsFgiR:-LADZ6Ca0C46gDAsFgiS:biejyke) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L68)):*
> ```Python
>         # All these cases are considered a "success" for the sandbox (it's the
>         # user's fault), but compilation is unsuccessful (no executable).
>         for code, exit_status in iteritems({
> ```

A dict doesn't seem a "natural" encoding here: why not keep it as a list of pairs?

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 72 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADYUzP2MY6GXT0GP2K:-LADYUzP2MY6GXT0GP2L:b-qorsmb) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L72)):*
> ```Python
>             "TO": Sandbox.EXIT_TIMEOUT,
>             "SG": Sandbox.EXIT_SIGNAL,
>         }):
> ```

Weird indentation...

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 73 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADYnue1YBodQQSMkiH:-LADYnue1YBodQQSMkiI:bgj5kra) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L73)):*
> ```Python
>             "SG": Sandbox.EXIT_SIGNAL,
>         }):
>             # A clean sandbox every time.
> ```

So, maybe, make three methods out of it? Or, at least, add a custom message to the asserts below to know in which case they failed.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 84 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZhKf1O0r9pTNBQZH:-LADZhKf1O0r9pTNBQZI:bjqmubl) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L84)):*
> ```Python
>             self.assertTrue(success)
>             self.assertFalse(compilation_success)
>             self.assertEqual(stats["exit_status"], exit_status)
> ```

Are the rest of the stats dropped when the compilation is unsuccessful? Why not check them?

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 105 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZyBc9GKiRdmIHqib:-LADZyBc9GKiRdmIHqic:b-sypj6z) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L105)):*
> ```Python
> 
>         success, compilation_success, text, stats = compilation_step(
>             self.sandbox, [["test", "command" "1"], ["command2"]])
> ```

Missing a comma between `"command"` and `"1"`? (same below)

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 108 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAD_B2fEaZFgm8Qyiqe:-LAD_B2fEaZFgm8Qyiqf:b-brgw1d) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L108)):*
> ```Python
> 
>         # 2 commands executed, with exec_num 0 and 1
>         self.assertEquals(self.sandbox.exec_num, 1)
> ```

If the sandbox was a mock you could assert what calls exactly were done! (same below)

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 121 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAD_QNI8UuIyP0Ms39E:-LAD_QNI8UuIyP0Ms39F:b-dxkwh7) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L121)):*
> ```Python
>         self.assertEqual(stats["stdout"].split("\n")[-1], "o2")
>         self.assertEqual(stats["stderr"].split("\n")[0], "e1")
>         self.assertEqual(stats["stderr"].split("\n")[-1], "e2")
> ```

Could maybe be simplified as a regexp match? (i.e., `o1.*o2`)

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 172 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAD_rbA25RGLaE5uAEj:-LAD_rbA25RGLaE5uAEk:brhumzt) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L172)):*
> ```Python
>         self.assertTrue(success)
>         self.assertTrue(compilation_success)
>         # UTF-8 invalid parts are replaced with funny question marks.
> ```

Same here.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 34 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADI-kz2kC73UOwRz7t:-LADI-kz2kC73UOwRz7u:bfzh9xp) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L34)):*
> ```Python
> 
> 
> class FakeIsolateSandbox(IsolateSandbox):
> ```

Is it necessary to extend IsolateSandbox? Isn't a properly-designed mock better?

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 37 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADHslWBx0lIS1HtVGw:-LADHslWBx0lIS1HtVGx:bb3b0sq) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L37)):*
> ```Python
>     """Isolate, with some fakeness sprinkled on top.
> 
>     This fake redefine execute_without_std to skip running the command and
> ```

s/redefine/redefines/

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 47 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJHEq23qYDqFHRP-8:-LADJHEq23qYDqFHRP-9:bcg9dwn) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L47)):*
> ```Python
>         self._fake_files = {}
> 
>         self._fake_execute_data = []
> ```

Since later you're popping from the front make this a deque?

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 67 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADIX5s3DeDycSqM8s7:-LADIX5s3DeDycSqM8s8:bc1zzna) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L67)):*
> ```Python
>         if memory is not None:
>             logs.append("max-rss:%d" % memory)
>             logs.append("cg-mem:%d" % memory)
> ```

IIRC this means that memory needs to be given in KiB: this isn't obvious. Either change it so that it takes bytes or add a docstring. (I prefer the first option)

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 73 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADIsPn6ngOQHP_jciz:-LADIsPn6ngOQHP_jcj-:bxav3tx) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L73)):*
> ```Python
>             logs.append("exitsig:%s" % signal)
>         data["log"] = "\n".join(logs).encode("utf-8")
>         self._fake_execute_data.append(data)
> ```

Blank lines in this method are a bit weird...

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 91 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJMAiBGaBjXIBkrip:-LADJMAiBGaBjXIBkriq:b-sb2guy) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L91)):*
> ```Python
>         assert len(self._fake_execute_data) > 0
> 
>         # Required for the correct operation of the base class.
> ```

Exactly what I was worried about in the above comment

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 31 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJx-_0NBBXoNVrr0l:-LADJx-_0NBBXoNVrr0m:btb0m4p) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L31)):*
> ```Python
> import unittest
> 
> from cmstestsuite.unit_tests.grading.steps.fakeisolatesandbox \
> ```

Shouldn't this be at the end of the cms* imports? Or, like DatabaseMixin, is there a reason to import it early?

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 68 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADKd-a3mKoBx2PIfoy:-LADKd-a3mKoBx2PIfoz:b5difyc) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L68)):*
> ```Python
>         self.sandbox.execute_without_std(["command"], wait=True)
> 
>         stats = execution_stats(self.sandbox)
> ```

Any reason to define this as a variable instead of inlining it in the assertEqual call? (also below)

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 38 at r3](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADLwPH0eoJaopBu5Pu:-LADLwPH0eoJaopBu5Pv:bvcxl87) ([raw file](https://github.com/cms-dev/cms/blob/9c5649a33b079a2ad6e44e62e0c9f27bb1a4a0ba/cmstestsuite/unit_tests/grading/steps/stats_test.py#L38)):*
> ```Python
> 
> 
> INVALID_UTF8 = b"\xc3\x28"
> ```

I think this isn't used anywhere (yet?). Maybe add it in the commit where it becomes necessary...

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 93 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADPnIq7nS4UK0iZz85:-LADPnIq7nS4UK0iZz86:brhumzt) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cmstestsuite/unit_tests/grading/steps/stats_test.py#L93)):*
> ```Python
> 
>         stats = execution_stats(self.sandbox, collect_output=True)
>         self.assertEqual(stats,
> ```

Same here.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 108 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADQ8tkD2lQMZ20ORSU:-LADQ8tkD2lQMZ20ORSV:b6rf8hn) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cmstestsuite/unit_tests/grading/steps/stats_test.py#L108)):*
> ```Python
>         stats = execution_stats(self.sandbox, collect_output=True)
> 
>         # UTF-8 invalid parts are replaced with funny question marks.
> ```

It's `\uFFFD` or `\N{REPLACEMENT CHARACTER}`, if you want to be more precise.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 198 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADQOHE3ThBHdUEXTAC:-LADQOHE3ThBHdUEXTAD:be2y968) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cmstestsuite/unit_tests/grading/steps/stats_test.py#L198)):*
> ```Python
>         r0 = get_stats(0, 0, 0, Sandbox.EXIT_OK, stdout="o1", stderr="e1")
>         r1 = get_stats(0, 0, 0, Sandbox.EXIT_OK, stdout="o2", stderr="e2")
>         m = merge_execution_stats(r0, r1)
> ```

Here too you could inline the call actually...

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[cms/grading/steps/stats.py, line 63 at r3](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADLKDNADREIiBobSRJ:-LADbya34-GqHLWPaVV5:bqt63gh) ([raw file](https://github.com/cms-dev/cms/blob/9c5649a33b079a2ad6e44e62e0c9f27bb1a4a0ba/cms/grading/steps/stats.py#L63)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Could be made a bit more general by saying that if either is None we return the other, and if both are None we return None. Though I'm not saying I prefer it this way, just wanting to point this out.
</blockquote></details>

Considered that, but the two inputs are ordered (see the logic to pick the first non-ok status). Having the first None means "merge the new one with nothing", having the second None means there is some error somewhere. 

---

*[cms/grading/steps/stats.py, line 117 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOdrK6tKauUWi3pIR:-LADcQNR5TcrncATf1Bj:blbk3tf) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L117)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Doing the join this way (pairwise) brings the overall complexity to N^2, whereas before it was N.
</blockquote></details>

I'm ok paying N^2 when N = 2 :P
Seriously, we should probably make this stat object more an object, preserving all information and then producing required stats on demand.

---

*[cms/grading/steps/stats.py, line 119 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOm698nrYwoTRCr59:-LADcuQ8E1131f0SDfbQ:b-uvtntd) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L119)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

I'm not sure I like this. I'm imagining the following scenario: three commands, all produce stdout but only the last one produces stderr. The merged stdout will be `a\n===\nb\n===\nc`, the stderr will be `foo`. It's impossible to figure out which command produced the stderr. If it were instead `\n===\n\n===\nfoo` it would be uglier but clear.
</blockquote></details>

This is even a mellower version of the original (forgot to say: this was a slight change in behaviour), and I changed it exactly to address your concern.

Original: keep only non-empty output/error and join them with "==="

Mine: keep empty output/error, but don't populate when neither of the original stats have that key.

The thing to consider is that any reasonable code would fill both stdout and stderr or none, and would do the same for all commands in the step.

I'm against what you're asking because at this point stdout and stderr are optional keys, and if we don't delete we are introducing them in the merge, giving the idea that output was recorded but was empty.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 63 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADRTz33kRKUBjWtoMD:-LAE1-jq0qGD0uroN5y1:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L63)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Did you consider importing `get_stats` from `stats_test` (or copying it over) to simplify this and all the ones below?
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 68 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZ6Ca0C46gDAsFgiR:-LAE1-Lt6S73gfQ8mJCl:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L68)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

A dict doesn't seem a "natural" encoding here: why not keep it as a list of pairs?
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 72 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADYUzP2MY6GXT0GP2K:-LAE10Lo-LwYBt0Gmf_1:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L72)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Weird indentation...
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 73 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADYnue1YBodQQSMkiH:-LAE1Z55B3YWIwYI1uQS:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L73)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

So, maybe, make three methods out of it? Or, at least, add a custom message to the asserts below to know in which case they failed.
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 84 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZhKf1O0r9pTNBQZH:-LAECKYoDcfFWHtLWjHi:b-ta3kdn) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L84)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Are the rest of the stats dropped when the compilation is unsuccessful? Why not check them?
</blockquote></details>

No they are not. Because it's not the point of this test, but if it makes feel you better, I'll check everything.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 105 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZyBc9GKiRdmIHqib:-LAECkWB0KdKOWFDQ6T4:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L105)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Missing a comma between `"command"` and `"1"`? (same below)
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 121 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAD_QNI8UuIyP0Ms39E:-LAECrzO05CrIF2aBxje:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L121)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Could maybe be simplified as a regexp match? (i.e., `o1.*o2`)
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 172 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAD_rbA25RGLaE5uAEj:-LAECy0uDplI3vnYhsDV:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L172)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 34 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADI-kz2kC73UOwRz7t:-LADveRoD88nhO3yBck1:b-khziix) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L34)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Is it necessary to extend IsolateSandbox? Isn't a properly-designed mock better?
</blockquote></details>

Maybe. I liked that this gives us a concise way to get all the stats (by writing a fake run.log file) without mocking every single call, plus reusing some of the isolate sandbox stuff makes it more reliable in terms of returning the correct types etc etc.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 37 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADHslWBx0lIS1HtVGw:-LADw1841q7thzhBVTyg:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L37)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

s/redefine/redefines/
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 47 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJHEq23qYDqFHRP-8:-LADw4iQ3UaBnLbCMHPy:bc9gbmr) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L47)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Since later you're popping from the front make this a deque?
</blockquote></details>

It's going to be so fast!

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 67 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADIX5s3DeDycSqM8s7:-LADy0Kh0VNRAW3LHMtq:b-nwaccq) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L67)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

IIRC this means that memory needs to be given in KiB: this isn't obvious. Either change it so that it takes bytes or add a docstring. (I prefer the first option)
</blockquote></details>

Added docstring. Prefer to keep in KiB as it is the format isolate reports.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 73 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADIsPn6ngOQHP_jciz:-LADyDFM6fc7fAhwFOfg:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L73)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Blank lines in this method are a bit weird...
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 91 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJMAiBGaBjXIBkrip:-LADyG0hBnipUgOswjv7:b-pjze3s) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L91)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Exactly what I was worried about in the above comment
</blockquote></details>

Nothing to be worried about IMHO - we are redefining this method, it's ok if some other parts of the class rely on behaviour that was present in the original method.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 31 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJx-_0NBBXoNVrr0l:-LADyXwlDTCQq-A1PCNV:br8kv6f) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L31)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Shouldn't this be at the end of the cms* imports? Or, like DatabaseMixin, is there a reason to import it early?
</blockquote></details>

We are in the cmstestsuite package, I assumed cmstestsuite imports go before other cms*. I honestly don't care though.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 68 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADKd-a3mKoBx2PIfoy:-LADyjds4k6H5jR3T4Gr:b9xscy0) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L68)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Any reason to define this as a variable instead of inlining it in the assertEqual call? (also below)
</blockquote></details>

Just usual setup - call - assert pattern

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 38 at r3](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADLwPH0eoJaopBu5Pu:-LADygy34aPkBo7tDwxP:bhy9ot6) ([raw file](https://github.com/cms-dev/cms/blob/9c5649a33b079a2ad6e44e62e0c9f27bb1a4a0ba/cmstestsuite/unit_tests/grading/steps/stats_test.py#L38)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

I think this isn't used anywhere (yet?). Maybe add it in the commit where it becomes necessary...
</blockquote></details>

Will do.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 108 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADQ8tkD2lQMZ20ORSU:-LAEDXjy82_yf8OenOoq:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cmstestsuite/unit_tests/grading/steps/stats_test.py#L108)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

It's `\uFFFD` or `\N{REPLACEMENT CHARACTER}`, if you want to be more precise.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940)*
<!-- Sent from Reviewable.io -->

lerks commented:
Reviewed 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cms/grading/steps/stats.py, line 117 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOdrK6tKauUWi3pIR:-LAFJwqf00SBaAOK9FVf:buy9o1j) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L117)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

I'm ok paying N^2 when N = 2 :P
Seriously, we should probably make this stat object more an object, preserving all information and then producing required stats on demand.
</blockquote></details>

Ok to keep it like this for now but I prefer the idea of the object as you suggested. (I prefer even more the possibility of storing them separately on the DB).

---

*[cms/grading/steps/stats.py, line 119 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOm698nrYwoTRCr59:-LAFOPfv9C9ZElyBBx4K:b-nk7k6h) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L119)):*
> keep empty output/error, but don't populate when neither of the original stats have that key.

I don't think the code does that. Merging a stats that has no stderr with one that has an empty one gives a stat that has no stderr. If you fix this I believe I will be happy. Also, since we disagreed on what the code does, maybe add a test please?

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 105 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZyBc9GKiRdmIHqib:-LAFNAZH6wRQPkF8xaZ0:b-9pg6gw) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L105)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can't see it

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 65 at r7](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAFL1fv5CdfEJUt-uop:-LAFL1fv5CdfEJUt-uoq:b-e8n09x) ([raw file](https://github.com/cms-dev/cms/blob/7dfd7b5ddb9fc9c1d3e9bfa01f1ff3cf0646dccc/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L65)):*
> ```Python
>         time (float): CPU time in seconds.
>         memory (int): memory used in KiB.
>         exit_status (string): isolate's two-letter exit status.
> ```

s/string/str/

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 31 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJx-_0NBBXoNVrr0l:-LAFNS2T65s1qvt_95Mv:b3y1l5b) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L31)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

We are in the cmstestsuite package, I assumed cmstestsuite imports go before other cms*. I honestly don't care though.
</blockquote></details>

I think all the absolute imports for cms* stuff should be in the same block properly sorted, but we've been inconsistent in this so you could keep it this way and then maybe one day we'll fix it everywhere with a global sweep.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
PTAL

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cms/grading/steps/stats.py, line 117 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOdrK6tKauUWi3pIR:-LAHYh5h1Youyj88bgJi:b-lvc3nh) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L117)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Ok to keep it like this for now but I prefer the idea of the object as you suggested. (I prefer even more the possibility of storing them separately on the DB).
</blockquote></details>

I'll write a TODO

---

*[cms/grading/steps/stats.py, line 119 at r4](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADOm698nrYwoTRCr59:-LAH_07aBYmCVrVkEAUB:b-fpeo2o) ([raw file](https://github.com/cms-dev/cms/blob/f949426777316b81660c9abd75b9e632dd22955a/cms/grading/steps/stats.py#L119)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

> keep empty output/error, but don't populate when neither of the original stats have that key.

I don't think the code does that. Merging a stats that has no stderr with one that has an empty one gives a stat that has no stderr. If you fix this I believe I will be happy. Also, since we disagreed on what the code does, maybe add a test please?
</blockquote></details>

Ah, you're right, because no === are introduced.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 105 at r5](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADZyBc9GKiRdmIHqib:-LAHZNQP8gjb7vs4h_Jc:b-pgglu7) ([raw file](https://github.com/cms-dev/cms/blob/644e390b11c1552d884bc45df58546d0fceed394/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L105)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Can't see it
</blockquote></details>

Oops, missed one.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 53 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC3-r9-53:-LAHZ2-p3kMUmnIpuiJR:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L53)):*
<details><summary><i>Previously, codacy-bot (Codacy Bot) wrote…</i></summary><blockquote>

![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'text'](https://app.codacy.com/app/s-maggiolo/cms/pullRequest?prid=1544702)
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 65 at r7](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAFL1fv5CdfEJUt-uop:-LAHZYukCazPI9k_5vaA:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/7dfd7b5ddb9fc9c1d3e9bfa01f1ff3cf0646dccc/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L65)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

s/string/str/
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 92 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC0-r9-92:-LAHZZgM6OQZxSqXCsvc:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L92)):*
<details><summary><i>Previously, codacy-bot (Codacy Bot) wrote…</i></summary><blockquote>

![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameters differ from overridden 'get_file' method](https://app.codacy.com/app/s-maggiolo/cms/pullRequest?prid=1544702)
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 31 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LADJx-_0NBBXoNVrr0l:-LAHZmFp5mZlhcJe-tR8:bfnnpbs) ([raw file](https://github.com/cms-dev/cms/blob/8dd90be7a6fd90b00fdba22f8be0f167900e2311/cmstestsuite/unit_tests/grading/steps/stats_test.py#L31)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

I think all the absolute imports for cms* stuff should be in the same block properly sorted, but we've been inconsistent in this so you could keep it this way and then maybe one day we'll fix it everywhere with a global sweep.
</blockquote></details>

Ok, sure. Changed here and in compilation_test

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940#-:-LAHaVlnAXGTKE5ut6Uv:bb71pro)*
<!-- Sent from Reviewable.io -->

lerks commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 2 of 2 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 53 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC3-r9-53:-LAIANvEAiYUzREWF-2i:bby2y7c) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L53)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not sure what you've "done" here exactly (there's no diff on these lines since last time I took a look and codacy keeps complaining)

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 92 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC0-r9-92:-LAIAaQq1lxdXwXO_JyU:b-v3ffra) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L92)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same here.

PS: if you just wanted to resolve codacy's comments you can do so by clicking on codacy's card in the bottom right deck and then selecting "dismiss from this discussion"; I just did that for both its new comments.

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 224 at r14](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAIAAPZEGf-23907tX1:-LAIAAPZEGf-23907tX2:blpcqxl) ([raw file](https://github.com/cms-dev/cms/blob/0b8b35727d840073648f60c154ec9f50cddb0cf8/cmstestsuite/unit_tests/grading/steps/stats_test.py#L224)):*
> ```Python
>             m, get_stats(0, 0, 0, Sandbox.EXIT_OK,
>                          stdout="o1\n===\n", stderr="\n===\ne2"))
> ```

Maybe also check that missing outputs are not preserved, up to you

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940#-:-LAIB1fk7iliUsXudp1c:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
Thanks

---

Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[cmstestsuite/unit_tests/grading/steps/compilation_test.py, line 53 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC3-r9-53:-LAIljF43oskq65s63Rd:b-huy8wu) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/compilation_test.py#L53)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Not sure what you've "done" here exactly (there's no diff on these lines since last time I took a look and codacy keeps complaining)
</blockquote></details>

Yeah, I wanted to mute it and "acknowledge doesn't make reviewable stop cycling through them. But they keep popping up...

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 92 at r9](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC0-r9-92:-LAIlzjECEN89M6Dxrpo:b-sk8x3j) ([raw file](https://github.com/cms-dev/cms/blob/5b8565d07c996d0ab9c3f8aab676fd144448d82f/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L92)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Same here.

PS: if you just wanted to resolve codacy's comments you can do so by clicking on codacy's card in the bottom right deck and then selecting "dismiss from this discussion"; I just did that for both its new comments.
</blockquote></details>

Ah thanks.

---

*[cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py, line 92 at r14](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LACksl4XTgBqTgaesC0-r14-92:-LAIm34r7yXn7iBvh6Sa:b1wku8q) ([raw file](https://github.com/cms-dev/cms/blob/0b8b35727d840073648f60c154ec9f50cddb0cf8/cmstestsuite/unit_tests/grading/steps/fakeisolatesandbox.py#L92)):*
<details><summary><i>Previously, codacy-bot (Codacy Bot) wrote…</i></summary><blockquote>

![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameters differ from overridden 'get_file' method](https://app.codacy.com/app/s-maggiolo/cms/pullRequest?prid=1544702)
</blockquote></details>

Not doing it 

---

*[cmstestsuite/unit_tests/grading/steps/stats_test.py, line 224 at r14](https://beta.reviewable.io/reviews/cms-dev/cms/940#-LAIAAPZEGf-23907tX1:-LAImr8b8YAbWNMxU3SJ:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/0b8b35727d840073648f60c154ec9f50cddb0cf8/cmstestsuite/unit_tests/grading/steps/stats_test.py#L224)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Maybe also check that missing outputs are not preserved, up to you
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940#-:-LAImsTJ9GCj-QH6jryt:b-zcjvte)*
<!-- Sent from Reviewable.io -->

lerks commented:
Reviewed 1 of 1 files at r15.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/940)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
Merged
